File: cpp/106_ConstructBinaryTreeFromInorderAndPostorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return build(inorder, postorder, INT_MAX);
    }
    
    TreeNode* build(vector<int>& inorder, vector<int>& postorder, int stop) {
        if (postorder.empty() || inorder[inorder.size()-1] == stop)
            return NULL;
        
        TreeNode* root = new TreeNode(postorder.back());
        postorder.pop_back();
        
        root->right = build(inorder, postorder, root->val);
        inorder.pop_back();
        root->left = build(inorder, postorder, stop);
        return root;
    }
};

File: python/106_ConstructBinaryTreeFromInorderAndPostorderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        def build(stop):
            if postorder and inorder[-1] != stop:
                root = TreeNode(postorder.pop())
                root.right = build(root.val)
                inorder.pop()
                root.left = build(stop)
                return root
        
        return build(None)
